libs/utils/android/benchmark.py

```diff
@@ -0,0 +1,262 @@
+        if self.bm_kind is None:
+            raise NotImplementedError("Override `bm_kind` attribute")
+        # Get a referench to the worload to run
+        self.wl = Workload(self.te).getInstance(self.bm_kind)
+        if self.wl is None:
+            raise ValueError("Specified benchmark [{}] is not supported"\
+                             .format(self.bm_kind))
+        return self.wl
+
+    def _getBmParams(self):
+        if self.bm_params is None:
+            raise NotImplementedError("Override `bm_params` attribute")
+        return self.bm_params
+
+    def _getBmCollect(self):
+        if self.bm_collect is None:
+            self.log.warning('No data collection configured')
+            return ''
+        return self.bm_collect
+
+    def __init__(self):
+        """
+        Set up logging and trigger running experiments
+        """
+        LisaLogging.setup()
+        self.log = logging.getLogger('Benchmark')
+
+        self.log.info('=== CommandLine parsing...')
+        self._parseCommandLine()
+
+        self.log.info('=== TestEnv setup...')
+        self.te = TestEnv(self._getBmConf())
+        self.target = self.te.target
+        # TODO: Check if TestEnv already provides support for rebooting
+        self._reboot()
+
+        self.log.info('=== Initialization...')
+        self._getBmKind()
+        self.out_dir=self.te.res_dir
+        if not self.benchmarkInit():
+            self.log.debug('Initialization failed: execution aborted')
+            return
+
+        self.log.info('=== Execution...')
+        self.wl.run(out_dir=self.out_dir,
+                    collect=self._getBmCollect(),
+                    **self.bm_params)
+
+        self.log.info('=== Finalization...')
+        self.benchmarkFinalize()
+
+    def _adb(self, cmd):
+        os.system('adb -s {} {}'.format(self.target.adb_name, cmd))
+
+    def _fastboot(self, cmd):
+        os.system('fastboot -s {} {}'.format(self.target.adb_name, cmd))
+
+
+    def _wait_for_logcat_idle(self, seconds=1):
+        lines = 0
+
+        # Clear logcat
+        # os.system('{} logcat -s {} -c'.format(adb, DEVICE));
+        self._adb('logcat -c')
+
+        # Dump logcat output
+        logcat_cmd = 'adb -s {} logcat'.format(self.target.adb_name)
+        logcat = Popen(logcat_cmd, shell=True, stdout=PIPE)
+        logcat_poll = select.poll()
+        logcat_poll.register(logcat.stdout, select.POLLIN)
+
+        # Monitor logcat until it's idle for the specified number of [s]
+        self.log.info('Waiting for system to be almost idle')
+        self.log.info('   i.e. at least %d[s] of no logcat messages', seconds)
+        while True:
+            poll_result = logcat_poll.poll(seconds * 1000)
+            if not poll_result:
+                break
+            lines = lines + 1
+            line = logcat.stdout.readline(1024)
+
+    def _reboot(self):
+
+        # Reboot the device, if a boot_image has been specified
+        if self.args.boot_image:
```

Comment:
Usually has never been a big issue.. however, there is a timeout in case of silent logcat.
I'm adding an additional "warn and return" in case of more than 1e6 logcat messages collected